pkg/controller/master/upgrade/upgrade_controller.go

```diff
@@ -869,120 +854,111 @@ func (h *upgradeHandler) setAutoCleanupSystemGeneratedSnapshotValue(value string
 	return nil
 }
 
-// storeVMState stores VM state only in single-node environments.
-// this method stores information of VMs with Ready=true (i.e. running state) in the upgrade annotation
-func (h *upgradeHandler) storeVMState(upgrade *harvesterv1.Upgrade) error {
+// restoreVM attempts to restore all running non-migratable VMs after the node upgrade succeeds.
+// During the node upgrade, in a multi-node environment, annotation harvesterhci.io/restore-vm-{node_uid}={vm_names_separated_by_comma}
+// will be added to upgrade CR for each node pre-drain job.
+// In a single-node environment, this will occur during the single-node-upgrade job.
+// Once the node upgrade is complete, VMs will be started based on the annotation.
+// Note that this function waits until KubeVirt is ready before restoring the VMs.
+// If KubeVirt is not ready, it will retry every 5 seconds by enqueuing the upgrade.
+func (h *upgradeHandler) restoreVM(upgrade *harvesterv1.Upgrade) error {
 	logFields := logrus.Fields{
 		"namespace": upgrade.Namespace,
 		"name":      upgrade.Name,
 	}
-	isSingleNodeRestoreVM, err := h.isSingleNodeRestoreVM(upgrade)
+
+	restoreVM, err := util.IsRestoreVM()
 	if err != nil {
-		return err
-	}
-	if !isSingleNodeRestoreVM {
-		logrus.WithFields(logFields).Info("Skip store VM state")
-		return nil
+		logrus.WithFields(logFields).WithError(err).Errorf("Failed to get setting UpgradeConfig, restoreVM fallback to default: %v", restoreVM)
 	}
 
-	if upgrade.Annotations != nil && upgrade.Annotations[preUpgradeVMsAnnotation] != "" {
-		logrus.WithFields(logFields).Infof("Skip store VM state since %s is already set", preUpgradeVMsAnnotation)
+	if !restoreVM {
 		return nil
 	}
 
-	vms, err := h.vmCache.List(corev1.NamespaceAll, labels.Everything())
+	err = h.checkKubeVirtHealth()
```

Comment:
Checking the AllComponentsReady condition in KubeVirt object is not sufficient, as VM start failures can still occur with the error 'the server is currently unable to handle the request.' occasionally. A more reliable approach is to call the `/healthz` API.